// polymer-linter/lib/html/undefined-elements.js
/**
 * @license
 * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const polymer_analyzer_1 = require("polymer-analyzer");
const registry_1 = require("../registry");
const rule_1 = require("./rule");
const stripIndent = require("strip-indent");
class UndefinedElements extends rule_1.HtmlRule {
    constructor() {
        super();
        this.code = 'undefined-elements';
        this.description = stripIndent(`
    Warns when an HTML tag refers to a custom element with no known definition.
  `).trim();
    }
    checkDocument(parsedDocument, document) {
        return __awaiter(this, void 0, void 0, function* () {
            const warnings = [];
            const refs = document.getByKind('element-reference');
            for (const ref of refs) {
                if (ref.tagName === 'test-fixture') {
                    // HACK. Filed as https://github.com/Polymer/polymer-analyzer/issues/507
                    continue;
                }
                const el = document.getById('element', ref.tagName, { imported: true, externalPackages: true });
                if (el.size === 0) {
                    warnings.push({
                        code: 'undefined-elements',
                        message: `The element ${ref.tagName} is not defined`,
                        severity: polymer_analyzer_1.Severity.WARNING,
                        sourceRange: parsedDocument.sourceRangeForStartTag(ref.astNode)
                    });
                }
            }
            return warnings;
        });
    }
}
exports.UndefinedElements = UndefinedElements;
registry_1.registry.register(new UndefinedElements());
